### PR TITLE
Application System: FileUploadField updates

### PR DESCRIPTION
**Pull request by madebynoam**
Updates and fixes to the FileUploadField

1. Fix 400 errors from AWS
2. Catch non 200 responses
3. Move the core of the field to shared, so that it can be used in a review screen
4. Add work in progress readme.

## TODO (next PR)
1. Add functionality to block the Continue button when uploading (currently the user can make it so that the answer is not saved if they click Continue before the upload is done).
2. Localise error messages (when Jeremy introduces the application system messages file in his PR).
3. See remaining items in readme for future updates

## Checklist:

- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] Formatting passes locally with my changes
- [x] I have rebased against main before asking for a review

### CODE REVIEW

File: libs/application/ui-fields/README.md [hunk 1]
```diff
@@ -9,3 +9,112 @@ Each Field has to have its own unique `type` which needs to be part of the `Fiel
...
+
+## FileUploadFormField
+
+The file upload form field provides a general way to upload files for your application.
+
+### To use the FileUploadFormField in an application
+
+1. Add it to the _schema_ using the following structure:
+
+```
```
Comment (jeremybarbet) [inline libs/application/ui-fields/README.md:23]: You can add `typescript` to get nice color syntax for the following code snippets

```suggestion
```typescript
```

File: libs/shared/form-fields/src/lib/FileUploadController/index.tsx [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+import { useLocale } from '@island.is/localization'
+
+type UploadFileAnswer = {
+  name: string
+  key?: string
+  url?: string
+}
+
+// Transform an uploaded file to an form answer.
+const transformToAnswer = (f: UploadFile): UploadFileAnswer => {
```
Comment (jeremybarbet) [inline libs/shared/form-fields/src/lib/FileUploadController/index.tsx:28]: You could destructurate to make it easier for the return

```suggestion
const transformToAnswer = (({ name, key, url }): UploadFile): UploadFileAnswer => {
  return { name, key, url }
```

File: libs/shared/form-fields/src/lib/FileUploadController/index.tsx [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+            key: fields.key,
+            url: `${response.url}/${fields.key}`,
+          },
+        },
+      })
+
+      // Done!
+      return Promise.resolve({ url: response.url, key: fields.key })
+    } catch (e) {
+      error = e
```
Comment (jeremybarbet) [inline libs/shared/form-fields/src/lib/FileUploadController/index.tsx:143]: Wouldn't it be better to have a useState for that with the default value to be the error props and then we can use the setter to pass the new e object as an error?

File: libs/shared/form-fields/src/lib/FileUploadController/index.tsx [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+    // with a spinner.
+    dispatch({
+      type: ActionTypes.ADD,
+      payload: {
+        newFiles: newUploadFiles,
+      },
+    })
+
+    // Upload each file.
+    newUploadFiles.forEach(async (f: UploadFile) => {
```
Comment (jeremybarbet) [inline libs/shared/form-fields/src/lib/FileUploadController/index.tsx:176]: Using async/await and then/catch together is a bit weird. I would probably do

```typescript
Promise.all(newUploadFiles.map(async (f: UploadFile) => {
  try {
    const res = await uploadFileFlow(f);
    
    dispatch({
      type: ActionTypes.UPDATE,
      payload: {
        file: f,
        status: 'done',
        percent: 100,
        key: res.key,
        url: res.url,
      },
    })
  } catch (e) {
    ...
  }
}))
```

File: libs/application/ui-fields/README.md (hunk 1 above)
Comment (madebynoam) [reply libs/application/ui-fields/README.md]: Oh nice thanks! Updated now.

File: libs/shared/form-fields/src/lib/FileUploadController/index.tsx (hunk 2 above)
Comment (madebynoam) [reply libs/shared/form-fields/src/lib/FileUploadController/index.tsx]: Nice, agree, switched now.

File: libs/shared/form-fields/src/lib/FileUploadController/index.tsx (hunk 4 above)
Comment (madebynoam) [reply libs/shared/form-fields/src/lib/FileUploadController/index.tsx:179]: Good point thanks, updated now.

File: libs/shared/form-fields/src/lib/FileUploadController/index.tsx (hunk 3 above)
Comment (madebynoam) [reply libs/shared/form-fields/src/lib/FileUploadController/index.tsx]: Makes sense thanks, I've refactored it to use the `uploadError` state consistently - with the `error` prop as the default.

**Comment by madebynoam** [comment]
Thanks for the feedback, should be ready for re-review.

**Review by jeremybarbet — APPROVED** [review]

**Pull request merged**